File: integration-tests-java/src/main/java/io/micronaut/jimmer/java/it/service/impl/BookImpl.java
package io.micronaut.jimmer.java.it.service.impl;

import static io.micronaut.jimmer.java.it.entity.Tables.BOOK_TABLE;

import io.micronaut.jimmer.java.it.entity.*;
import io.micronaut.jimmer.java.it.service.IBook;
import jakarta.inject.Singleton;
import jakarta.transaction.Transactional;
import java.math.BigDecimal;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.ast.mutation.SimpleSaveResult;
import org.babyfish.jimmer.sql.ast.tuple.Tuple2;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.jetbrains.annotations.Nullable;

@Singleton
public class BookImpl implements IBook {

    BookTable table = BOOK_TABLE;

    private final JSqlClient sqlClient;

    public BookImpl(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
    }

    @Override
    public Book findById(long id) {
        return sqlClient.findById(Book.class, id);
    }

    @Override
    public Book findById(int id, Fetcher<Book> fetcher) {
        return sqlClient.findById(fetcher, id);
    }

    @Override
    public List<Book> findByIds(List<Integer> ids) {
        return sqlClient.findByIds(Book.class, ids);
    }

    @Override
    public SimpleSaveResult<Book> save(Book book) {
        SimpleSaveResult<Book> save = sqlClient.save(book);
        int i = 1 / 0;
        return save;
    }

    @Override
    public Map<Long, BigDecimal> findAvgPriceGroupByStoreId(Collection<Long> storeIds) {
        return Tuple2.toMap(
                sqlClient
                        .createQuery(table)
                        .where(table.storeId().in(storeIds))
                        .groupBy(table.storeId())
                        .select(table.storeId(), table.price().avgAsDecimal())
                        .execute());
    }

    @Override
    public List<Book> findBooksByName(@Nullable String name, @Nullable Fetcher<Book> fetcher) {
        return sqlClient
                .createQuery(table)
                .whereIf(name != null && !name.isEmpty(), table.name().ilike(name))
                .select(table.fetch(fetcher))
                .execute();
    }

    @Override
    public List<Book> findBooksByName(@Nullable String name) {
        return sqlClient
                .createQuery(table)
                .whereIf(name != null && !name.isEmpty(), table.name().ilike(name))
                .select(
                        table.fetch(
                                Fetchers.BOOK_FETCHER
                                        .allScalarFields()
                                        .store(Fetchers.BOOK_STORE_FETCHER.name())))
                .execute();
    }

    @Override
    @Transactional
    public void update() {
        sqlClient
                .createUpdate(Tables.BOOK_STORE_TABLE)
                .set(Tables.BOOK_STORE_TABLE.website(), "https://www.manning.com")
                .where(Tables.BOOK_STORE_TABLE.id().eq(2L))
                .execute();
    }

    @Override
    public List<Book> manyToMany() {
        return sqlClient
                .createQuery(table)
                .where(table.edition().eq(1))
                .select(
                        table.fetch(
                                Fetchers.BOOK_FETCHER
                                        .allScalarFields()
                                        .authors(Fetchers.AUTHOR_FETCHER.allScalarFields())))
                .execute();
    }

    @Override
    public void updateOneToMany() {
        sqlClient
                .createUpdate(table)
                .set(table.store().id(), 2L)
                .where(table.id().eq(7L))
                .execute();
    }

    @Override
    public void saveManyToMany() {
        sqlClient.getAssociations(BookProps.AUTHORS).save(10, 3);
    }
}

File: integration-tests-kotlin/src/main/kotlin/io/micronaut/jimmer/kotlin/it/service/impl/BookImpl.kt
package io.micronaut.jimmer.kotlin.it.service.impl

import io.micronaut.jimmer.kotlin.it.entity.Book
import io.micronaut.jimmer.kotlin.it.entity.BookStore
import io.micronaut.jimmer.kotlin.it.entity.by
import io.micronaut.jimmer.kotlin.it.entity.edition
import io.micronaut.jimmer.kotlin.it.entity.id
import io.micronaut.jimmer.kotlin.it.entity.name
import io.micronaut.jimmer.kotlin.it.entity.price
import io.micronaut.jimmer.kotlin.it.entity.store
import io.micronaut.jimmer.kotlin.it.entity.website
import io.micronaut.jimmer.kotlin.it.service.IBook
import io.micronaut.transaction.annotation.Transactional
import jakarta.inject.Singleton
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.asNonNull
import org.babyfish.jimmer.sql.kt.ast.expression.avgAsDecimal
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.ast.expression.like
import org.babyfish.jimmer.sql.kt.ast.expression.valueIn
import org.babyfish.jimmer.sql.kt.ast.mutation.KSimpleSaveResult
import org.babyfish.jimmer.sql.kt.ast.query.whereIfNotBlank
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import java.math.BigDecimal

@Singleton
open class BookImpl(
    private val sqlClient: KSqlClient,
) : IBook {
    override fun findById(id: Long): Book? = sqlClient.findById(Book::class, id)

    override fun findById(
        id: Int,
        fetcher: Fetcher<Book>,
    ): Book? = sqlClient.findById(fetcher, id)

    override fun findByIds(ids: List<Int>): List<Book> = sqlClient.findByIds(Book::class, ids)

    @Transactional(rollbackFor = [Exception::class])
    override fun save(book: Book): KSimpleSaveResult<Book> {
        val save = sqlClient.save(book)
        val i = 1 / 0
        return save
    }

    override fun findAvgPriceGroupByStoreId(storeIds: Collection<Long>): Map<Long, BigDecimal> =
        sqlClient
            .createQuery(Book::class) {
                where(table.store.id valueIn storeIds)
                groupBy(table.store.id)
                select(
                    table.store.id,
                    avgAsDecimal(table.price).asNonNull(),
                )
            }.execute()
            .associateBy({ it._1 }) {
                it._2
            }

    override fun findBooksByName(
        name: String,
        fetcher: Fetcher<Book>,
    ): List<Book> =
        sqlClient
            .createQuery(Book::class) {
                whereIfNotBlank(name) {
                    table.name.like(name)
                }
                select(table.fetch(fetcher))
            }.execute()

    override fun findBooksByName(name: String): List<Book> =
        sqlClient
            .createQuery(Book::class) {
                whereIfNotBlank(name) {
                    table.name.like(name)
                }
                select(
                    table.fetch(
                        newFetcher(Book::class).by {
                            allScalarFields()
                            store { name }
                        },
                    ),
                )
            }.execute()

    @Transactional(rollbackFor = [Exception::class])
    override fun update() {
        sqlClient
            .createUpdate(BookStore::class) {
                set(table.website, "https://www.manning.com")
                where(table.id eq 2L)
            }.execute()
    }

    override fun manyToMany(): List<Book> =
        sqlClient
            .createQuery(Book::class) {
                where(table.edition eq 1)
                select(
                    table.fetch(
                        newFetcher(Book::class).by {
                            allScalarFields()
                            authors { allScalarFields() }
                            store { allScalarFields() }
                        },
                    ),
                )
            }.execute()

    @Transactional(rollbackFor = [Exception::class])
    override fun updateOneToMany() {
        sqlClient
            .createUpdate(Book::class) {
                set(table.store.id, 2L)
                where(table.id eq 7L)
            }.execute()
    }

    @Transactional(rollbackFor = [Exception::class])
    override fun saveManyToMany() {
        sqlClient.getAssociations(Book::authors).save(10, 3)
    }
}
